## Full Training Run

In [1]:
from sklearn import logger

from pyspark.sql import SparkSession
from utils.dataPipeline import DataPreparation, connect_spark_session
from utils.ALSModelPipeline import ALSModelPipeline

config = {
    "base_date_str": "2026-05-10",
    "db_path": "gs://wynk-ml-workspace/projects/xstream_nlu/catalog-db/",
    "temp_path": "gs://wynk-ml-workspace/_temp/harshith/als",
    "test_days": 30,
    "train_days": 90,
    "playtime_history_path": "gs://wynk-ml-workspace/projects/rails_reranking/daily_user_watch_history_new",
    "click_history_path": "gs://wynk-ml-workspace/projects/rails_reranking/daily_user_click_watch_history_new",
    "distinct_user_content_threshold": 3
}

# ==========================================
# SCENARIO A: FULL TRAINING RUN
# ==========================================
spark = connect_spark_session()
# 1. Prep Data
decayed_data_prep = DataPreparation(spark, config, decay=True)
decayed_data_prep.prepare_features_and_metadata()


bash: /opt/conda/miniconda3/lib/libtinfo.so.6: no version information available (required by bash)
bash: /opt/conda/miniconda3/lib/libtinfo.so.6: no version information available (required by bash)
26/05/18 06:07:09 WARN SparkConf: The configuration key 'spark.yarn.executor.failuresValidityInterval' has been deprecated as of Spark 3.5 and may be removed in the future. Please use the new key 'spark.executor.failuresValidityInterval' instead.
26/05/18 06:07:10 WARN SparkConf: The configuration key 'spark.yarn.executor.failuresValidityInterval' has been deprecated as of Spark 3.5 and may be removed in the future. Please use the new key 'spark.executor.failuresValidityInterval' instead.
Setting default log level to "WARN".
To adjust logging level use sc.setLogLevel(newLevel). For SparkR, use setLogLevel(newLevel).
26/05/18 06:07:10 WARN SparkConf: The configuration key 'spark.yarn.executor.failuresValidityInterval' has been deprecated as of Spark 3.5 and may be removed in the future. Pleas

Connecting to Spark Session RecSysPipeline...


Path not found or inaccessible, skipping: gs://wynk-ml-workspace/projects/rails_reranking/daily_user_watch_history_new/day=2026-02-06
Path not found or inaccessible, skipping: gs://wynk-ml-workspace/projects/rails_reranking/daily_user_watch_history_new/day=2026-02-05


Data read from gs://wynk-ml-workspace/projects/rails_reranking/daily_user_watch_history_new completed, filtering for common users...
self train and test df have been assigned
Playtime combined successfully.
Playtime combined successfully.
Playtime to label completed.


ERROR:root:Exception while sending command.
Traceback (most recent call last):
  File "/opt/conda/default/lib/python3.11/site-packages/py4j/clientserver.py", line 511, in send_command
    answer = smart_decode(self.stream.readline()[:-1])
                          ^^^^^^^^^^^^^^^^^^^^^^
RuntimeError: reentrant call inside <_io.BufferedReader name=77>

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/opt/conda/default/lib/python3.11/site-packages/py4j/java_gateway.py", line 1038, in send_command
    response = connection.send_command(command)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/conda/default/lib/python3.11/site-packages/py4j/clientserver.py", line 539, in send_command
    raise Py4JNetworkError(
py4j.protocol.Py4JNetworkError: Error while sending or receiving
ERROR:root:Exception while sending command.
Traceback (most recent call last):
  File "/opt/conda/default/lib/python3.11/site-packages/py4j/cl

Unexpected error validating gs://wynk-ml-workspace/projects/rails_reranking/daily_user_click_watch_history_new/day=2026-02-17: An error occurred while calling o1234.parquet
Path not found or inaccessible, skipping: gs://wynk-ml-workspace/projects/rails_reranking/daily_user_click_watch_history_new/day=2026-02-06
Path not found or inaccessible, skipping: gs://wynk-ml-workspace/projects/rails_reranking/daily_user_click_watch_history_new/day=2026-02-05


Data read from gs://wynk-ml-workspace/projects/rails_reranking/daily_user_click_watch_history_new completed, filtering for common users...
Successfully combined training and testing features!


26/05/18 06:08:31 WARN SparkStringUtils: Truncated the string representation of a plan since it was too large. This behavior can be adjusted by setting 'spark.sql.debug.maxToStringFields'.


In [2]:
test_users = [
    {
        "userId": "User_1",
        "watch_history": "HOTSTAR_DTH_TVSHOW_1271337449, HOTSTAR_DTH_MOVIE_1271509950, HOTSTAR_DTH_MOVIE_1271305185"
    },
    {
        "userId": "User_2",
        "watch_history": "HOTSTAR_DTH_MOVIE_1271511297, HOTSTAR_DTH_MOVIE_1271527032, HOTSTAR_DTH_TVSHOW_1271419298"
    },
    {
        "userId": "User_3",
        "watch_history": ["SHEMAROOME_MOVIE_5f20071fa609d206c800aa88", "HOTSTAR_DTH_MOVIE_3595", "HOTSTAR_DTH_TVSHOW_1271461884", "HOTSTAR_DTH_MOVIE_1442", "HOTSTAR_DTH_MOVIE_1271406274", "HOTSTAR_DTH_MOVIE_1271435228", "SONYLIV_VOD_TVSHOW_1700000852" ] 
    }
]

decayed_data_prep.add_custom_user_data(test_users, confidence_score=4.0)
decayed_data_prep.build_indices_and_ground_truth(penalize_popularity= False)
# 3. Save ALL Checkpoints (Data + Predictions)
decayed_data_prep.save_data_checkpoint()

/home/B0336827/.local/lib/python3.11/site-packages/pandas/core/computation/expressions.py:22: UserWarning: Pandas requires version '2.10.2' or newer of 'numexpr' (version '2.10.0' currently installed).
  from pandas.core.computation.check import NUMEXPR_INSTALLED
/home/B0336827/.local/lib/python3.11/site-packages/pandas/core/arrays/masked.py:56: UserWarning: Pandas requires version '1.4.2' or newer of 'bottleneck' (version '1.3.8' currently installed).
  from pandas.core import (


Successfully added 3 custom users to the training data.
Filtering users, aggregating playtime, and building indexes...
Applied time decay to confidence scores.


26/05/13 05:54:07 WARN SparkConf: The configuration key 'spark.yarn.executor.failuresValidityInterval' has been deprecated as of Spark 3.5 and may be removed in the future. Please use the new key 'spark.executor.failuresValidityInterval' instead.


Generating Ground Truth...
Saving data checkpoint to: default_checkpoint...
  Saving indexed_test...


26/05/13 05:55:18 WARN SparkConf: The configuration key 'spark.yarn.executor.failuresValidityInterval' has been deprecated as of Spark 3.5 and may be removed in the future. Please use the new key 'spark.executor.failuresValidityInterval' instead.
26/05/13 05:55:19 WARN SparkConf: The configuration key 'spark.yarn.executor.failuresValidityInterval' has been deprecated as of Spark 3.5 and may be removed in the future. Please use the new key 'spark.executor.failuresValidityInterval' instead.
26/05/13 05:55:19 WARN SparkConf: The configuration key 'spark.yarn.executor.failuresValidityInterval' has been deprecated as of Spark 3.5 and may be removed in the future. Please use the new key 'spark.executor.failuresValidityInterval' instead.
26/05/13 05:55:26 WARN SparkConf: The configuration key 'spark.yarn.executor.failuresValidityInterval' has been deprecated as of Spark 3.5 and may be removed in the future. Please use the new key 'spark.executor.failuresValidityInterval' instead.
26/05/13 05:

  Saving indexed_train...


26/05/13 05:55:45 WARN SparkConf: The configuration key 'spark.yarn.executor.failuresValidityInterval' has been deprecated as of Spark 3.5 and may be removed in the future. Please use the new key 'spark.executor.failuresValidityInterval' instead.
26/05/13 05:55:46 WARN SparkConf: The configuration key 'spark.yarn.executor.failuresValidityInterval' has been deprecated as of Spark 3.5 and may be removed in the future. Please use the new key 'spark.executor.failuresValidityInterval' instead.
26/05/13 05:56:05 WARN SparkConf: The configuration key 'spark.yarn.executor.failuresValidityInterval' has been deprecated as of Spark 3.5 and may be removed in the future. Please use the new key 'spark.executor.failuresValidityInterval' instead.
26/05/13 05:56:06 WARN SparkConf: The configuration key 'spark.yarn.executor.failuresValidityInterval' has been deprecated as of Spark 3.5 and may be removed in the future. Please use the new key 'spark.executor.failuresValidityInterval' instead.


  Saving ground_data...


  Saving item_lookup...
  Saving user_lookup...


  Saving combined_train_df...


  Saving combined_test_df...


26/05/13 05:56:46 WARN SparkConf: The configuration key 'spark.yarn.executor.failuresValidityInterval' has been deprecated as of Spark 3.5 and may be removed in the future. Please use the new key 'spark.executor.failuresValidityInterval' instead.


  Saving metadata_df...


  Saving item_popularity...


Checkpoint successfully saved!


In [3]:
config = {
    "base_date_str": "2026-03-28",
    "db_path": "gs://wynk-ml-workspace/projects/xstream_nlu/catalog-db/",
    "temp_path": "gs://wynk-ml-workspace/_temp/harshith/als",
    "test_days": 30,
    "train_days": 90,
    "playtime_history_path": "gs://wynk-ml-workspace/projects/rails_reranking/daily_user_watch_history_new",
    "click_history_path": "gs://wynk-ml-workspace/projects/rails_reranking/daily_user_click_watch_history_new",
    "distinct_user_content_threshold": 3
}
fresh_data_prep = DataPreparation(spark, config)
fresh_data_prep.load_data_checkpoint()
# 2. Train Model
als_pipeline = ALSModelPipeline(spark, config, data_prep=fresh_data_prep)
als_params = {
    "maxIter": 15,
    "coldStartStrategy": "drop",
    "rank": 50,          # Drop from 100 to 50 to prevent overfitting sparse data
    "alpha": 10.0,       # Give the implicit signals a bit more weight
    "k": 50,             # Generate more raw recs to filter later
    "implicitPrefs": True,
    "regParam": 0.05     # <-- CRITICAL FIX. Drop this from 1.0.
}
als_pipeline.train_ALS(als_params, print_metrics= True)

# 3. Save ALL Checkpoints (Data + Predictions)
als_pipeline.save_model_checkpoint()

# 4. Test Recs
# als_pipeline.get_recommendations_userId("ib-xPv06AjuYOE4BI0")

Loading data checkpoint from: gs://wynk-ml-workspace/_temp/harshith/als/default_checkpoint...
Checkpoint loaded successfully! Ready for training or inference.
Training ALS Model...


Generating 50 raw recommendations for all users...
Filtering raw recommendations down to 50 novel items per user...
Applying post-processing popularity penalty to ALS scores...

EVALUATING MODEL METRICS


26/05/13 06:02:27 WARN SparkConf: The configuration key 'spark.yarn.executor.failuresValidityInterval' has been deprecated as of Spark 3.5 and may be removed in the future. Please use the new key 'spark.executor.failuresValidityInterval' instead.
26/05/13 06:04:08 WARN SparkConf: The configuration key 'spark.yarn.executor.failuresValidityInterval' has been deprecated as of Spark 3.5 and may be removed in the future. Please use the new key 'spark.executor.failuresValidityInterval' instead.
26/05/13 06:04:08 WARN SparkConf: The configuration key 'spark.yarn.executor.failuresValidityInterval' has been deprecated as of Spark 3.5 and may be removed in the future. Please use the new key 'spark.executor.failuresValidityInterval' instead.
26/05/13 06:04:13 WARN SparkConf: The configuration key 'spark.yarn.executor.failuresValidityInterval' has been deprecated as of Spark 3.5 and may be removed in the future. Please use the new key 'spark.executor.failuresValidityInterval' instead.
26/05/13 06:

Catalog Coverage (@50): 0.0776  (2711 / 34921 items recommended)

[Ranking Performance]


26/05/13 06:04:22 WARN SparkConf: The configuration key 'spark.yarn.executor.failuresValidityInterval' has been deprecated as of Spark 3.5 and may be removed in the future. Please use the new key 'spark.executor.failuresValidityInterval' instead.
26/05/13 06:04:24 WARN SparkConf: The configuration key 'spark.yarn.executor.failuresValidityInterval' has been deprecated as of Spark 3.5 and may be removed in the future. Please use the new key 'spark.executor.failuresValidityInterval' instead.


Users Evaluated (Train/Test intersection): 2304266
Precision@50: 0.0186
Recall@50:    0.1856
MAP:             0.0426

Saving ALL ALS artifacts to gs://wynk-ml-workspace/_temp/harshith/als/als_checkpoint...


Model checkpoint successfully saved!


# Testing Code

In [3]:
config = {
    "base_date_str": "2026-04-12",
    "db_path": "gs://wynk-ml-workspace/projects/xstream_nlu/catalog-db/",
    "temp_path": "gs://wynk-ml-workspace/_temp/harshith/als",
    "test_days": 30,
    "train_days": 90,
    "playtime_history_path": "gs://wynk-ml-workspace/projects/rails_reranking/daily_user_watch_history_new",
    "click_history_path": "gs://wynk-ml-workspace/projects/rails_reranking/daily_user_click_watch_history_new",
    "distinct_user_content_threshold": 3
}

In [8]:
import pyspark.sql.functions as F
# catalog = spark.read.parquet("gs://wynk-ml-workspace/projects/xstream_nlu/catalog-db/")
# catalog.printSchema()

base_db_path = f"{config['db_path']}{config['base_date_str']}"
tv_df = spark.read.parquet(f"{base_db_path}/enriched_tv.parquet")
movie_df = spark.read.parquet(f"{base_db_path}/enriched_movie.parquet")

def clean_meta(df):
    return df.filter((F.col('XstreamContentIds') == F.array())) \
        .withColumn("item_id_exploded", F.explode("XstreamContentIds")) \
        .select(
            F.col("item_id_exploded").cast("string").alias("item_id"),
            "title", 
            F.col('OriginalLanguage').alias('original_language').cast("string"),
            "Genres"
        )
            
metadata_df = clean_meta(tv_df).unionByName(clean_meta(movie_df)).distinct()

In [12]:
clean_meta(tv_df).filter(F.col("item_id") == "HOICHOI_TVSHOW_f93c5116-3f0b-44dc-bef4-811ed4e4ea76").show(truncate=False)

+-------+-----+-----------------+------+
|item_id|title|original_language|Genres|
+-------+-----+-----------------+------+
+-------+-----+-----------------+------+



In [22]:
playtime_history = f"{config['playtime_history_path']}/{config['base_date_str']}"
spark.read.parquet(playtime_history).filter(F.col("userId") == "User_3").show(truncate=False)


AnalysisException: [PATH_NOT_FOUND] Path does not exist: gs://wynk-ml-workspace/projects/rails_reranking/daily_user_watch_history_new/2026-04-12.

# Load and Run

In [1]:
from utils.dataPipeline import DataPreparation, connect_spark_session
from utils.ALSModelPipeline import ALSModelPipeline

config = {
    "base_date_str": "2026-03-28",
    "db_path": "gs://wynk-ml-workspace/projects/xstream_nlu/catalog-db/",
    "temp_path": "gs://wynk-ml-workspace/_temp/harshith/als",
    "test_days": 30,
    "train_days": 90,
    "playtime_history_path": "gs://wynk-ml-workspace/projects/rails_reranking/daily_user_watch_history_new",
    "click_history_path": "gs://wynk-ml-workspace/projects/rails_reranking/daily_user_click_watch_history_new",
    "distinct_user_content_threshold": 1
}
spark = connect_spark_session()
fresh_data_prep = DataPreparation(spark, config, True)
fresh_data_prep.load_data_checkpoint()
# 2. Load Model Checkpoint (Restores pre-computed als_top_k)
als_pipeline = ALSModelPipeline(spark, config, data_prep=fresh_data_prep)
als_pipeline.load_model_checkpoint()

# 3. Serve Recs Instantly
# fresh_als_pipeline.get_recommendations_userId("ib-xPv06AjuYOE4BI0")

bash: /opt/conda/miniconda3/lib/libtinfo.so.6: no version information available (required by bash)
bash: /opt/conda/miniconda3/lib/libtinfo.so.6: no version information available (required by bash)
26/05/22 06:20:07 WARN SparkConf: The configuration key 'spark.yarn.executor.failuresValidityInterval' has been deprecated as of Spark 3.5 and may be removed in the future. Please use the new key 'spark.executor.failuresValidityInterval' instead.
26/05/22 06:20:07 WARN SparkConf: The configuration key 'spark.yarn.executor.failuresValidityInterval' has been deprecated as of Spark 3.5 and may be removed in the future. Please use the new key 'spark.executor.failuresValidityInterval' instead.
Setting default log level to "WARN".
To adjust logging level use sc.setLogLevel(newLevel). For SparkR, use setLogLevel(newLevel).
26/05/22 06:20:08 WARN SparkConf: The configuration key 'spark.yarn.executor.failuresValidityInterval' has been deprecated as of Spark 3.5 and may be removed in the future. Pleas

Connecting to Spark Session RecSysPipeline...
Loading data checkpoint from: gs://wynk-ml-workspace/_temp/harshith/als/default_checkpoint...


Checkpoint loaded successfully! Ready for training or inference.
Loading ALS artifacts from: gs://wynk-ml-workspace/_temp/harshith/als/als_checkpoint...
Model checkpoint loaded successfully! Ready for inference.


In [2]:
fresh_data_prep.print_data_funnel_metrics()


OMEGA-LEVEL MATRIX DIAGNOSTICS: TOTAL PIPELINE EXPOSURE


26/05/22 06:02:19 WARN CacheManager: Asked to cache already cached data.



[1. BASE DIMENSIONS & MATRIX SHAPE]
  Filled Matrix Cells:     41,761,518
  Unique Users (U):        2,412,861
  Unique Items (I):        34,921
  Exact Sparsity:          99.950437%
  Matrix Shape Factor:     69.09 Users for every 1 Item



[2. USER ENGAGEMENT SKEW]
  Distributions (Items per User):
    Min/25th/Median:       3 / 6 / 11
    75th/90th/95th:        22 / 37 / 49
    99th / Max (Whale):    3940 / 3940
  Anomalies:
    Zombie Users (=1 rec): 0 (0.0%)
    Cold Users (<5 recs):  380,073 (15.8%)
    Binge Quotient:        Max user watched 11.28% of entire catalog

[3. ITEM POPULARITY SKEW]
  Distributions (Users per Item):
    Min/25th/Median:       1 / 10 / 43
    75th/90th/95th:        256 / 1297 / 3133
    99th / Max (Hit):      497230 / 497230
  Anomalies & Wealth Skew:
    One-Hit Wonders:       1,756 (5.0%)
    Cold Items (<5 users): 4,110 (11.8%)
    Gini Proxy:            Top 10% of items hoard 89.7% of all traffic.
                           Bottom 50% of items share only 0.6% of traffic.

[4. IMPLICIT SIGNAL VARIANCE & DEGRADATION]
  Confidence Min / Max:    0.1000 / 4.0000
  Confidence Mean:         1.5908
  Confidence StdDev:       1.0983 (Must be > 0.2 to matter)
  Decay Graveyard (<=0.1): 87,818 ro

  Score ~ 0.1:  3,581,364 (  8.6%)
  Score ~ 0.2:  4,370,933 ( 10.5%)
  Score ~ 0.6:  1,191,930 (  2.9%)
  Score ~ 0.7:  1,358,117 (  3.3%)
  Score ~ 0.8:  1,503,510 (  3.6%)
  Score ~ 0.9:  1,778,044 (  4.3%)
  Score ~ 1.0:  1,512,076 (  3.6%)
  Score ~ 1.2:    989,102 (  2.4%)
  Score ~ 1.5:  1,117,279 (  2.7%)
  Score ~ 1.6:  1,543,993 (  3.7%)
  Score ~ 1.7:  1,613,957 (  3.9%)
  Score ~ 1.8:  1,791,750 (  4.3%)
  Score ~ 1.9:  1,625,028 (  3.9%)
  Score ~ 2.0:  1,402,137 (  3.4%)
  Score ~ 2.9:  1,098,876 (  2.6%)




In [ ]:
als_pipeline.print_omega_model_diagnostics()


OMEGA-LEVEL MODEL DIAGNOSTICS: BEHAVIOR & OUTPUT SKEW



[1. CATALOG COVERAGE & UTILIZATION]
  Users Evaluated:         2,412,859
  Total Recs Generated:    96,935,497
  Unique Items Suggested:  2,670 out of 34,921 (7.65% Coverage)



[2. RECOMMENDATION SKEW (Diversity of Output)]
  Median Item Rec'd:       9856 times
  95th Percentile:         143158 times
  99th Percentile:         520502 times
  Max (The Model's Fav):   520502 times
  -> WARNING ALARM: The top 1% of items consume 0.5% of all recommendation slots.



[3. POPULARITY BIAS AMPLIFICATION]
  Avg Popularity Score (1-100) of Recommended Items: 12.69
  (Model is successfully surfacing middle-tier content.)



[4. CATALOG RESURRECTION (The Niche Test)]
  Out of the bottom 17,291 niche items in training...
  The model recommended 0 of them at least once (0.0% Resurrection Rate).




26/05/22 06:22:51 WARN YarnSchedulerBackend$YarnSchedulerEndpoint: Requesting driver to remove executor 5 for reason Executor for container container_1764236692086_9308_01_000005 exited because of a YARN event (e.g., preemption) and not because of an error in the running job.
26/05/22 06:22:51 WARN YarnSchedulerBackend$YarnSchedulerEndpoint: Requesting driver to remove executor 6 for reason Executor for container container_1764236692086_9308_01_000006 exited because of a YARN event (e.g., preemption) and not because of an error in the running job.


In [2]:
als_pipeline.evaluate_model()


EVALUATING MODEL METRICS



--- Recommendation Frequency Percentiles ---
25th Percentile: 1.1 rec_frequency_score(1-100)
50th Percentile: 2.2 rec_frequency_score(1-100)
75th Percentile: 6.3 rec_frequency_score(1-100)
90th Percentile: 13.8 rec_frequency_score(1-100)
95th Percentile: 22.4 rec_frequency_score(1-100)
99th Percentile: 49.2 rec_frequency_score(1-100)


/usr/lib/spark/python/pyspark/sql/context.py:158: FutureWarning: Deprecated in 3.0.0. Use SparkSession.builder.getOrCreate() instead.
  warnings.warn(
/home/B0336827/.local/lib/python3.11/site-packages/pandas/core/computation/expressions.py:22: UserWarning: Pandas requires version '2.10.2' or newer of 'numexpr' (version '2.10.0' currently installed).
  from pandas.core.computation.check import NUMEXPR_INSTALLED
/home/B0336827/.local/lib/python3.11/site-packages/pandas/core/arrays/masked.py:56: UserWarning: Pandas requires version '1.4.2' or newer of 'bottleneck' (version '1.3.8' currently installed).
  from pandas.core import (


Catalog Coverage (@50): 0.0776  (2711 / 34921 items recommended)

[Ranking Performance]


Users Evaluated (Train/Test intersection): 2304266
Precision@15: 0.0362
Recall@15:    0.1856
MAP:             0.0426



{'coverage': 0.07763237020703874,
 'avg_recall': 0.1855550411105965,
 'precision_k': 0.03621914020921775,
 'map': 0.042588349051999014}

26/05/18 08:09:50 WARN YarnSchedulerBackend$YarnSchedulerEndpoint: Requesting driver to remove executor 5 for reason Executor for container container_1764236692086_9042_01_000007 exited because of a YARN event (e.g., preemption) and not because of an error in the running job.
26/05/18 08:09:50 WARN YarnSchedulerBackend$YarnSchedulerEndpoint: Requesting driver to remove executor 9 for reason Executor for container container_1764236692086_9042_01_000012 exited because of a YARN event (e.g., preemption) and not because of an error in the running job.
26/05/18 08:09:50 WARN YarnSchedulerBackend$YarnSchedulerEndpoint: Requesting driver to remove executor 6 for reason Executor for container container_1764236692086_9042_01_000009 exited because of a YARN event (e.g., preemption) and not because of an error in the running job.
26/05/18 08:09:50 WARN YarnSchedulerBackend$YarnSchedulerEndpoint: Requesting driver to remove executor 8 for reason Executor for container container_1764236692086_9042_

## Generate recommendations

User_1 (Superhero-- Daredevil, Superman): HOTSTAR_DTH_TVSHOW_1271337449, HOTSTAR_DTH_TVSHOW_1271337449

User_2 (Horror-- Dies Irae, RajaSaab, IT): HOTSTAR_DTH_MOVIE_1271511297, HOTSTAR_DTH_MOVIE_1271527032, HOTSTAR_DTH_TVSHOW_1271419298

User_3 (Mix-- welcome, home alone, AKOTSK, Blr Days, Tron, black bag, rocket boys): SHEMAROOME_MOVIE_5f20071fa609d206c800aa88, HOTSTAR_DTH_MOVIE_3595, HOTSTAR_DTH_TVSHOW_1271461884, HOTSTAR_DTH_MOVIE_1442, HOTSTAR_DTH_MOVIE_1271406274, HOTSTAR_DTH_MOVIE_1271435228, SONYLIV_VOD_TVSHOW_1700000852 

In [5]:
als_pipeline.get_recommendations_userId("User_1")
als_pipeline.get_recommendations_userId("User_2")
als_pipeline.get_recommendations_userId("User_3")


Watch/Click History for userId User_1:


26/05/13 06:04:37 WARN SparkConf: The configuration key 'spark.yarn.executor.failuresValidityInterval' has been deprecated as of Spark 3.5 and may be removed in the future. Please use the new key 'spark.executor.failuresValidityInterval' instead.
26/05/13 06:04:38 WARN SparkConf: The configuration key 'spark.yarn.executor.failuresValidityInterval' has been deprecated as of Spark 3.5 and may be removed in the future. Please use the new key 'spark.executor.failuresValidityInterval' instead.


Watch History Count:  3


26/05/13 06:04:39 WARN SparkConf: The configuration key 'spark.yarn.executor.failuresValidityInterval' has been deprecated as of Spark 3.5 and may be removed in the future. Please use the new key 'spark.executor.failuresValidityInterval' instead.


+---------------------+-----------------------------+------------------------------------+-----------------+
|title                |item_id                      |Genres                              |original_language|
+---------------------+-----------------------------+------------------------------------+-----------------+
|Deadpool & Wolverine |HOTSTAR_DTH_MOVIE_1271305185 |[Action, Comedy, Science Fiction]   |en               |
|Daredevil: Born Again|HOTSTAR_DTH_TVSHOW_1271337449|[Drama, Action & Adventure, Crime]  |en               |
|Superman             |HOTSTAR_DTH_MOVIE_1271509950 |[Science Fiction, Adventure, Action]|en               |
+---------------------+-----------------------------+------------------------------------+-----------------+


Top Novel Recommendations for userId User_1:


26/05/13 06:04:42 WARN SparkConf: The configuration key 'spark.yarn.executor.failuresValidityInterval' has been deprecated as of Spark 3.5 and may be removed in the future. Please use the new key 'spark.executor.failuresValidityInterval' instead.
26/05/13 06:04:42 WARN SparkConf: The configuration key 'spark.yarn.executor.failuresValidityInterval' has been deprecated as of Spark 3.5 and may be removed in the future. Please use the new key 'spark.executor.failuresValidityInterval' instead.
26/05/13 06:04:42 WARN SparkConf: The configuration key 'spark.yarn.executor.failuresValidityInterval' has been deprecated as of Spark 3.5 and may be removed in the future. Please use the new key 'spark.executor.failuresValidityInterval' instead.


+------------------------------+---------------------------------------------+-----------------+----------------+-----------------------------+
|title                         |Genres                                       |original_language|popularity_score|item_id                      |
+------------------------------+---------------------------------------------+-----------------+----------------+-----------------------------+
|The Raja Saab                 |[Horror, Comedy, Romance]                    |te               |10              |HOTSTAR_DTH_MOVIE_1271527032 |
|Kaattaan                      |[Drama, Crime, Action & Adventure]           |ta               |8               |HOTSTAR_DTH_TVSHOW_1271512941|
|The Conjuring: Last Rites     |[Horror]                                     |en               |6               |HOTSTAR_DTH_MOVIE_1271526246 |
|Sarvam Maya                   |[Comedy, Family]                             |ml               |6               |HOTSTAR_DTH_MOVIE_12715

26/05/13 06:04:43 WARN SparkConf: The configuration key 'spark.yarn.executor.failuresValidityInterval' has been deprecated as of Spark 3.5 and may be removed in the future. Please use the new key 'spark.executor.failuresValidityInterval' instead.
26/05/13 06:04:43 WARN SparkConf: The configuration key 'spark.yarn.executor.failuresValidityInterval' has been deprecated as of Spark 3.5 and may be removed in the future. Please use the new key 'spark.executor.failuresValidityInterval' instead.


Watch History Count:  3


26/05/13 06:04:44 WARN SparkConf: The configuration key 'spark.yarn.executor.failuresValidityInterval' has been deprecated as of Spark 3.5 and may be removed in the future. Please use the new key 'spark.executor.failuresValidityInterval' instead.


+--------------------+-----------------------------+-------------------------+-----------------+
|title               |item_id                      |Genres                   |original_language|
+--------------------+-----------------------------+-------------------------+-----------------+
|Diés Iraé           |HOTSTAR_DTH_MOVIE_1271511297 |[Horror, Thriller]       |ml               |
|IT: Welcome to Derry|HOTSTAR_DTH_TVSHOW_1271419298|[Drama, Mystery]         |en               |
|The Raja Saab       |HOTSTAR_DTH_MOVIE_1271527032 |[Horror, Comedy, Romance]|te               |
+--------------------+-----------------------------+-------------------------+-----------------+


Top Novel Recommendations for userId User_2:


26/05/13 06:04:46 WARN SparkConf: The configuration key 'spark.yarn.executor.failuresValidityInterval' has been deprecated as of Spark 3.5 and may be removed in the future. Please use the new key 'spark.executor.failuresValidityInterval' instead.
26/05/13 06:04:47 WARN SparkConf: The configuration key 'spark.yarn.executor.failuresValidityInterval' has been deprecated as of Spark 3.5 and may be removed in the future. Please use the new key 'spark.executor.failuresValidityInterval' instead.
26/05/13 06:04:47 WARN SparkConf: The configuration key 'spark.yarn.executor.failuresValidityInterval' has been deprecated as of Spark 3.5 and may be removed in the future. Please use the new key 'spark.executor.failuresValidityInterval' instead.


+------------------------------+-----------------------------------------------+-----------------+----------------+-----------------------------+
|title                         |Genres                                         |original_language|popularity_score|item_id                      |
+------------------------------+-----------------------------------------------+-----------------+----------------+-----------------------------+
|Kaattaan                      |[Drama, Crime, Action & Adventure]             |ta               |8               |HOTSTAR_DTH_TVSHOW_1271512941|
|The Conjuring: Last Rites     |[Horror]                                       |en               |6               |HOTSTAR_DTH_MOVIE_1271526246 |
|Sarvam Maya                   |[Comedy, Family]                               |ml               |6               |HOTSTAR_DTH_MOVIE_1271525191 |
|Chiraiya                      |[Drama]                                        |hi               |6               |HOTSTAR_D

26/05/13 06:04:47 WARN SparkConf: The configuration key 'spark.yarn.executor.failuresValidityInterval' has been deprecated as of Spark 3.5 and may be removed in the future. Please use the new key 'spark.executor.failuresValidityInterval' instead.
26/05/13 06:04:48 WARN SparkConf: The configuration key 'spark.yarn.executor.failuresValidityInterval' has been deprecated as of Spark 3.5 and may be removed in the future. Please use the new key 'spark.executor.failuresValidityInterval' instead.


Watch History Count:  7


26/05/13 06:04:49 WARN SparkConf: The configuration key 'spark.yarn.executor.failuresValidityInterval' has been deprecated as of Spark 3.5 and may be removed in the future. Please use the new key 'spark.executor.failuresValidityInterval' instead.


+------------------------------+-----------------------------------------+---------------------------------------------+-----------------+
|title                         |item_id                                  |Genres                                       |original_language|
+------------------------------+-----------------------------------------+---------------------------------------------+-----------------+
|Bangalore Days                |HOTSTAR_DTH_MOVIE_1442                   |[Drama, Family, Romance]                     |ml               |
|Home Alone                    |HOTSTAR_DTH_MOVIE_3595                   |[Comedy, Family]                             |en               |
|Tron: Ares                    |HOTSTAR_DTH_MOVIE_1271406274             |[Science Fiction, Adventure, Action]         |en               |
|Rocket Boys                   |SONYLIV_VOD_TVSHOW_1700000852            |[Drama]                                      |hi               |
|A Knight of the Seven King

26/05/13 06:04:51 WARN SparkConf: The configuration key 'spark.yarn.executor.failuresValidityInterval' has been deprecated as of Spark 3.5 and may be removed in the future. Please use the new key 'spark.executor.failuresValidityInterval' instead.
26/05/13 06:04:51 WARN SparkConf: The configuration key 'spark.yarn.executor.failuresValidityInterval' has been deprecated as of Spark 3.5 and may be removed in the future. Please use the new key 'spark.executor.failuresValidityInterval' instead.
26/05/13 06:04:51 WARN SparkConf: The configuration key 'spark.yarn.executor.failuresValidityInterval' has been deprecated as of Spark 3.5 and may be removed in the future. Please use the new key 'spark.executor.failuresValidityInterval' instead.


+----------------------------------+------------------------------------+-----------------+----------------+------------------------------------+
|title                             |Genres                              |original_language|popularity_score|item_id                             |
+----------------------------------+------------------------------------+-----------------+----------------+------------------------------------+
|Shark Tank India                  |[Reality]                           |hi               |31              |SONYLIV_VOD_TVSHOW_1700000741       |
|Freedom at Midnight               |[Drama, War & Politics]             |hi               |21              |SONYLIV_VOD_TVSHOW_1700001477       |
|Maharani                          |[Drama]                             |hi               |18              |SONYLIV_VOD_TVSHOW_1700000725       |
|Gullak                            |[Comedy, Drama, Family]             |hi               |14              |SONYLIV_VOD_TVSH

DataFrame[title: string, Genres: array<string>, original_language: string, popularity_score: int, item_id: string]

In [6]:
als_pipeline.get_recommendations_random_user(min_content_threshold=4)

26/05/13 06:04:53 WARN SparkConf: The configuration key 'spark.yarn.executor.failuresValidityInterval' has been deprecated as of Spark 3.5 and may be removed in the future. Please use the new key 'spark.executor.failuresValidityInterval' instead.



Randomly Selected User: k4eF9OLZulNo9A98q0
User's Historical Item Count: 7

Watch/Click History for userId k4eF9OLZulNo9A98q0:


26/05/13 06:04:54 WARN SparkConf: The configuration key 'spark.yarn.executor.failuresValidityInterval' has been deprecated as of Spark 3.5 and may be removed in the future. Please use the new key 'spark.executor.failuresValidityInterval' instead.


Watch History Count:  8


26/05/13 06:04:55 WARN SparkConf: The configuration key 'spark.yarn.executor.failuresValidityInterval' has been deprecated as of Spark 3.5 and may be removed in the future. Please use the new key 'spark.executor.failuresValidityInterval' instead.


+--------------------+---------------------------------------------------------------+-------------------------+-----------------+
|title               |item_id                                                        |Genres                   |original_language|
+--------------------+---------------------------------------------------------------+-------------------------+-----------------+
|Sankalp             |MINITV_TVSHOW_amzn1.dv.gti.9de167a6-7185-4687-a3ca-8ac4198143d2|[Drama]                  |hi               |
|Sankalp             |MINITV_TVSHOW_amzn1.dv.gti.9de167a6-7185-4687-a3ca-8ac4198143d2|[]                       |hi               |
|Mrs.                |ZEEFIVE_MOVIE_0-0-1z5689119                                    |[Drama]                  |hi               |
|Gullak              |SONYLIV_VOD_TVSHOW_1700000659                                  |[Comedy, Drama, Family]  |hi               |
|The Aam Aadmi Family|ZEEFIVE_TVSHOW_0-6-3351                                      

26/05/13 06:04:57 WARN SparkConf: The configuration key 'spark.yarn.executor.failuresValidityInterval' has been deprecated as of Spark 3.5 and may be removed in the future. Please use the new key 'spark.executor.failuresValidityInterval' instead.
26/05/13 06:04:57 WARN SparkConf: The configuration key 'spark.yarn.executor.failuresValidityInterval' has been deprecated as of Spark 3.5 and may be removed in the future. Please use the new key 'spark.executor.failuresValidityInterval' instead.
26/05/13 06:04:57 WARN SparkConf: The configuration key 'spark.yarn.executor.failuresValidityInterval' has been deprecated as of Spark 3.5 and may be removed in the future. Please use the new key 'spark.executor.failuresValidityInterval' instead.


+-------------------------------------+----------------------------------+-----------------+----------------+---------------------------------------------------------------+
|title                                |Genres                            |original_language|popularity_score|item_id                                                        |
+-------------------------------------+----------------------------------+-----------------+----------------+---------------------------------------------------------------+
|The Taj Story                        |[Drama]                           |hi               |45              |LIONSGATEPLAY_MOVIE_THETAJSTORYY2025MHI                        |
|The Kerala Story                     |[Drama]                           |hi               |43              |ZEEFIVE_MOVIE_0-0-1z5512788                                    |
|Sitaare Zameen Par                   |[Comedy, Drama]                   |hi               |38              |SONYLIV_VOD_MOVIE_109

DataFrame[title: string, Genres: array<string>, original_language: string, popularity_score: int, item_id: string]

In [7]:
als_pipeline.get_recommendations_userId("FdPTwN400Ax9YLEYD0")


Watch/Click History for userId FdPTwN400Ax9YLEYD0:


26/05/11 08:08:51 WARN SparkConf: The configuration key 'spark.yarn.executor.failuresValidityInterval' has been deprecated as of Spark 3.5 and may be removed in the future. Please use the new key 'spark.executor.failuresValidityInterval' instead.
26/05/11 08:08:53 WARN SparkConf: The configuration key 'spark.yarn.executor.failuresValidityInterval' has been deprecated as of Spark 3.5 and may be removed in the future. Please use the new key 'spark.executor.failuresValidityInterval' instead.


Watch History Count:  61


26/05/11 08:08:55 WARN SparkConf: The configuration key 'spark.yarn.executor.failuresValidityInterval' has been deprecated as of Spark 3.5 and may be removed in the future. Please use the new key 'spark.executor.failuresValidityInterval' instead.


+----------------------------------+---------------------------------------------------+----------------------------------------------+
|title                             |item_id                                            |Genres                                        |
+----------------------------------+---------------------------------------------------+----------------------------------------------+
|Greenland                         |LIONSGATEPLAY_MOVIE_GREENLANDY2020M                |[Action, Adventure, Science Fiction, Thriller]|
|Freedom at Midnight               |SONYLIV_VOD_TVSHOW_1700001477                      |[Drama, War & Politics]                       |
|Biriyaani                         |SHEMAROOME_MOVIE_627b9e664210bb32ba003753          |[Drama]                                       |
|Karma Korma                       |HOICHOI_TVSHOW_67f4bd8b-ee9f-4901-b71f-326257403273|[Drama]                                       |
|The Puppet’s Tale                 |HOICHOI_MOVI

26/05/11 08:08:58 WARN SparkConf: The configuration key 'spark.yarn.executor.failuresValidityInterval' has been deprecated as of Spark 3.5 and may be removed in the future. Please use the new key 'spark.executor.failuresValidityInterval' instead.
26/05/11 08:08:58 WARN SparkConf: The configuration key 'spark.yarn.executor.failuresValidityInterval' has been deprecated as of Spark 3.5 and may be removed in the future. Please use the new key 'spark.executor.failuresValidityInterval' instead.
26/05/11 08:08:58 WARN SparkConf: The configuration key 'spark.yarn.executor.failuresValidityInterval' has been deprecated as of Spark 3.5 and may be removed in the future. Please use the new key 'spark.executor.failuresValidityInterval' instead.


+------------------------------+--------------------------------------+--------------------+---------------------------------------------------+
|title                         |Genres                                |popularity_score    |item_id                                            |
+------------------------------+--------------------------------------+--------------------+---------------------------------------------------+
|Joto Kando Kolkatatei         |[Mystery, Drama]                      |0.05743446467577576 |ZEEFIVE_MOVIE_0-0-1z5888766                        |
|NULL                          |NULL                                  |0.05647119483431559 |ZEEFIVE_TVSHOW_0-6-4z5869712                       |
|NULL                          |NULL                                  |0.04372485655872849 |ZEEFIVE_MOVIE_0-0-1z5852916                        |
|Grihapravesh                  |[Drama, Family, Romance]              |0.03616858421316642 |HOICHOI_MOVIE_db24323b-4ddb-433b-8898-

DataFrame[title: string, Genres: array<string>, popularity_score: double, item_id: string]

1. Content Threshold = 5
2. popularity decay
3. regularisation

# Special cases testing

In [7]:
# Animesh
# als_pipeline.get_recommendations_userId("ib-xPv06AjuYOE4BI0")

# Narendra
als_pipeline.get_recommendations_userId("p-cWmUEqg4MGRdaJp0")

#Ajita
als_pipeline.get_recommendations_userId("YNkVt86cnYgIYKBOM0")

#Random
als_pipeline.get_recommendations_random_user(min_content_threshold=4)


Watch/Click History for userId p-cWmUEqg4MGRdaJp0:


26/05/13 06:11:42 WARN SparkConf: The configuration key 'spark.yarn.executor.failuresValidityInterval' has been deprecated as of Spark 3.5 and may be removed in the future. Please use the new key 'spark.executor.failuresValidityInterval' instead.
26/05/13 06:11:43 WARN SparkConf: The configuration key 'spark.yarn.executor.failuresValidityInterval' has been deprecated as of Spark 3.5 and may be removed in the future. Please use the new key 'spark.executor.failuresValidityInterval' instead.


Watch History Count:  26


26/05/13 06:11:44 WARN SparkConf: The configuration key 'spark.yarn.executor.failuresValidityInterval' has been deprecated as of Spark 3.5 and may be removed in the future. Please use the new key 'spark.executor.failuresValidityInterval' instead.


+-------------------------+----------------------------------------------+------------------------------------+-----------------+
|title                    |item_id                                       |Genres                              |original_language|
+-------------------------+----------------------------------------------+------------------------------------+-----------------+
|NULL                     |HOTSTAR_DTH_TVSHOW_1271524483                 |NULL                                |NULL             |
|Arrival                  |LIONSGATEPLAY_MOVIE_ARRIVALY2016M             |[Drama, Science Fiction, Mystery]   |en               |
|Geetha Govindam          |ZEEFIVE_MOVIE_0-0-7463                        |[Romance]                           |te               |
|Shambhala                |AHA_MOVIE_91B37D19-AF75-4184-884F-7C2EC2809368|[Action, Horror, Thriller]          |te               |
|14 Phere                 |ZEEFIVE_MOVIE_0-0-1z51198                     |[Comedy, Romance

26/05/13 06:11:47 WARN SparkConf: The configuration key 'spark.yarn.executor.failuresValidityInterval' has been deprecated as of Spark 3.5 and may be removed in the future. Please use the new key 'spark.executor.failuresValidityInterval' instead.
26/05/13 06:11:47 WARN SparkConf: The configuration key 'spark.yarn.executor.failuresValidityInterval' has been deprecated as of Spark 3.5 and may be removed in the future. Please use the new key 'spark.executor.failuresValidityInterval' instead.
26/05/13 06:11:47 WARN SparkConf: The configuration key 'spark.yarn.executor.failuresValidityInterval' has been deprecated as of Spark 3.5 and may be removed in the future. Please use the new key 'spark.executor.failuresValidityInterval' instead.


+-------------------------------------+----------------------------------------------+-----------------+----------------+----------------------------------------------+
|title                                |Genres                                        |original_language|popularity_score|item_id                                       |
+-------------------------------------+----------------------------------------------+-----------------+----------------+----------------------------------------------+
|Mana ShankaraVaraprasad Garu         |[Action, Comedy, Mystery]                     |te               |100             |ZEEFIVE_MOVIE_0-0-1z5904059                   |
|Bhartha Mahasayulaku Wignyapthi      |[Action, Family, Drama]                       |te               |56              |ZEEFIVE_MOVIE_0-0-1z5916299                   |
|Taarak Mehta Ka Ooltah Chashmah      |[Comedy, Drama, Family, Kids, Soap]           |hi               |53              |SONYLIV_VOD_TVSHOW_1700000084     

26/05/13 06:11:48 WARN SparkConf: The configuration key 'spark.yarn.executor.failuresValidityInterval' has been deprecated as of Spark 3.5 and may be removed in the future. Please use the new key 'spark.executor.failuresValidityInterval' instead.


  - [No title metadata found for history]
User ID YNkVt86cnYgIYKBOM0 not found in user lookup.



Randomly Selected User: i7kdXIYE1ESGo-Kfc0
User's Historical Item Count: 35

Watch/Click History for userId i7kdXIYE1ESGo-Kfc0:


26/05/13 06:11:52 WARN SparkConf: The configuration key 'spark.yarn.executor.failuresValidityInterval' has been deprecated as of Spark 3.5 and may be removed in the future. Please use the new key 'spark.executor.failuresValidityInterval' instead.
26/05/13 06:11:53 WARN SparkConf: The configuration key 'spark.yarn.executor.failuresValidityInterval' has been deprecated as of Spark 3.5 and may be removed in the future. Please use the new key 'spark.executor.failuresValidityInterval' instead.


Watch History Count:  35


26/05/13 06:11:54 WARN SparkConf: The configuration key 'spark.yarn.executor.failuresValidityInterval' has been deprecated as of Spark 3.5 and may be removed in the future. Please use the new key 'spark.executor.failuresValidityInterval' instead.


+-------------------------------+--------------------------------------------------------------+---------------------------------------------+-----------------+
|title                          |item_id                                                       |Genres                                       |original_language|
+-------------------------------+--------------------------------------------------------------+---------------------------------------------+-----------------+
|MasterChef India               |SONYLIV_VOD_TVSHOW_1700001090                                 |[Reality]                                    |hi               |
|Mane Lai Jaa                   |SHEMAROOME_MOVIE_63344a59dde7153e24000ae2                     |[Drama]                                      |gu               |
|Sarvam Maya                    |HOTSTAR_DTH_MOVIE_1271525191                                  |[Comedy, Family]                             |ml               |
|Bhairavam                      |Z

26/05/13 06:11:56 WARN SparkConf: The configuration key 'spark.yarn.executor.failuresValidityInterval' has been deprecated as of Spark 3.5 and may be removed in the future. Please use the new key 'spark.executor.failuresValidityInterval' instead.
26/05/13 06:11:56 WARN SparkConf: The configuration key 'spark.yarn.executor.failuresValidityInterval' has been deprecated as of Spark 3.5 and may be removed in the future. Please use the new key 'spark.executor.failuresValidityInterval' instead.
26/05/13 06:11:56 WARN SparkConf: The configuration key 'spark.yarn.executor.failuresValidityInterval' has been deprecated as of Spark 3.5 and may be removed in the future. Please use the new key 'spark.executor.failuresValidityInterval' instead.


+----------------------------------------+----------------------------------+-----------------+----------------+--------------------------------------------------------------+
|title                                   |Genres                            |original_language|popularity_score|item_id                                                       |
+----------------------------------------+----------------------------------+-----------------+----------------+--------------------------------------------------------------+
|Sirai                                   |[Drama, Romance]                  |ta               |84              |ZEEFIVE_MOVIE_0-0-1z5895270                                   |
|Parasakthi                              |[History, Drama, Action, Romance] |ta               |59              |ZEEFIVE_MOVIE_0-0-1z5902844                                   |
|Sitaare Zameen Par                      |[Comedy, Drama]                   |hi               |38              |SONYLIV_

DataFrame[title: string, Genres: array<string>, original_language: string, popularity_score: int, item_id: string]

In [9]:
als_pipeline.get_recommendations_random_user(min_content_threshold=4)

26/05/13 06:13:35 WARN SparkConf: The configuration key 'spark.yarn.executor.failuresValidityInterval' has been deprecated as of Spark 3.5 and may be removed in the future. Please use the new key 'spark.executor.failuresValidityInterval' instead.



Randomly Selected User: Mn0B8PTSci_V7j6HZ0
User's Historical Item Count: 8

Watch/Click History for userId Mn0B8PTSci_V7j6HZ0:


26/05/13 06:13:36 WARN SparkConf: The configuration key 'spark.yarn.executor.failuresValidityInterval' has been deprecated as of Spark 3.5 and may be removed in the future. Please use the new key 'spark.executor.failuresValidityInterval' instead.


Watch History Count:  8


26/05/13 06:13:37 WARN SparkConf: The configuration key 'spark.yarn.executor.failuresValidityInterval' has been deprecated as of Spark 3.5 and may be removed in the future. Please use the new key 'spark.executor.failuresValidityInterval' instead.


+--------------------+-----------------------------------------------+-------------------------------------------+-----------------+
|title               |item_id                                        |Genres                                     |original_language|
+--------------------+-----------------------------------------------+-------------------------------------------+-----------------+
|Ratri Ke Yatri2     |HUNGAMA_TVSHOW_61663384                        |[Drama]                                    |hi               |
|Ayalaan             |AHA_MOVIE_1C526B44-8389-4ED1-A5B1-0DF7F8433942 |[Science Fiction, Action, Adventure]       |ta               |
|96                  |AHA_MOVIE_1005A59D-1135-47CC-8D90-BC159AB46C23 |[Romance]                                  |ta               |
|C.I.D.              |SONYLIV_VOD_TVSHOW_1700000196                  |[Action & Adventure, Crime, Drama, Mystery]|hi               |
|Sulthan             |AHA_MOVIE_3F86D041-EBBB-4028-90EC-B199AC580FB9 

26/05/13 06:13:39 WARN SparkConf: The configuration key 'spark.yarn.executor.failuresValidityInterval' has been deprecated as of Spark 3.5 and may be removed in the future. Please use the new key 'spark.executor.failuresValidityInterval' instead.
26/05/13 06:13:39 WARN SparkConf: The configuration key 'spark.yarn.executor.failuresValidityInterval' has been deprecated as of Spark 3.5 and may be removed in the future. Please use the new key 'spark.executor.failuresValidityInterval' instead.
26/05/13 06:13:39 WARN SparkConf: The configuration key 'spark.yarn.executor.failuresValidityInterval' has been deprecated as of Spark 3.5 and may be removed in the future. Please use the new key 'spark.executor.failuresValidityInterval' instead.


+-------------------------+--------------------------------------------+-----------------+----------------+--------------------------------------------------------------+
|title                    |Genres                                      |original_language|popularity_score|item_id                                                       |
+-------------------------+--------------------------------------------+-----------------+----------------+--------------------------------------------------------------+
|Shambhala                |[Action, Horror, Thriller]                  |te               |41              |AHA_MOVIE_91B37D19-AF75-4184-884F-7C2EC2809368                |
|Om Shanti Shanti Shantihi|[]                                          |te               |31              |AHA_MOVIE_0DC150B0-1DDA-41F5-8971-51688F2EBA22                |
|Sarkaar                  |[Reality]                                   |te               |29              |AHA_TVSHOW_836A0A87-656C-46FE-BBD1-F12

DataFrame[title: string, Genres: array<string>, original_language: string, popularity_score: int, item_id: string]

In [10]:
from pyspark.sql import functions as F

item_popularity = fresh_data_prep.item_popularity.select("item_id", "popularity_score", "user_count","popularity_penalty")

# Define the percentiles you want (expressed as decimals 0.0 to 1.0)
percentiles = [0.25, 0.50, 0.75, 0.90, 0.95, 0.99]

# Calculate the values
# 0.01 is the relative error: lower is more accurate but slower
quantile_values = item_popularity.filter(F.col("popularity_score")>0).stat.approxQuantile(
    "popularity_score", 
    percentiles, 
    0.01
)

# Pair them up and print the results
print("--- Item Popularity Percentiles ---")
for p, v in zip(percentiles, quantile_values):
    print(f"{int(p*100)}th Percentile: {v} popularity_score(1-100)")

quantile_values = item_popularity.filter(F.col("popularity_score")>0).stat.approxQuantile(
    "user_count", 
    percentiles, 
    0.01
)
# Pair them up and print the results
print("\n\n--- Item Popularity Percentiles ---")
for p, v in zip(percentiles, quantile_values):
    print(f"{int(p*100)}th Percentile: {v} user_count")

quantile_values = item_popularity.filter(F.col("popularity_score")>0).stat.approxQuantile(
    "popularity_penalty", 
    percentiles, 
    0.01
)
# Pair them up and print the results
print("\n\n--- Item Popularity Percentiles ---")
for p, v in zip(percentiles, quantile_values):
    print(f"{int(p*100)}th Percentile: {v} popularity_penalty")

--- Item Popularity Percentiles ---
25th Percentile: 1.0 popularity_score(1-100)
50th Percentile: 1.0 popularity_score(1-100)
75th Percentile: 1.0 popularity_score(1-100)
90th Percentile: 1.0 popularity_score(1-100)
95th Percentile: 2.0 popularity_score(1-100)
99th Percentile: 100.0 popularity_score(1-100)


--- Item Popularity Percentiles ---
25th Percentile: 10.0 user_count
50th Percentile: 44.0 user_count
75th Percentile: 257.0 user_count
90th Percentile: 1307.0 user_count
95th Percentile: 3205.0 user_count
99th Percentile: 504720.0 user_count


--- Item Popularity Percentiles ---
25th Percentile: 0.4046466855875561 popularity_penalty
50th Percentile: 0.567077308316706 popularity_penalty
75th Percentile: 0.756304195516401 popularity_penalty
90th Percentile: 0.8725028695491559 popularity_penalty
95th Percentile: 0.9266284080291267 popularity_penalty
99th Percentile: 0.9602525677891276 popularity_penalty


Catalog Coverage (@50): 0.0574  (2325 / 40478 items recommended)

[Ranking Performance]Precision@15: 0.0406
Recall@15:    0.1103
MAP:             0.0390

# Implicit Prefs = False

Whever the reg Param > 1, the model is performing very poorly in all aspects, so skipping it diring hyp Param tuning`